integrations/gcp/gcp_core/search/searches.py

```diff
@@ -0,0 +1,186 @@
+    """
+    Search for projects that the caller has ``resourcemanager.projects.get`` permission on
+    """
+    async with ProjectsAsyncClient() as projects_client:
+        search_projects_pager = await projects_client.search_projects()
+        async for projects_page in search_projects_pager.pages:
+            yield parse_protobuf_messages(projects_page.projects)
+
+
+async def search_all_folders() -> ASYNC_GENERATOR_RESYNC_TYPE:
+    """
+    Search for folders that the caller has ``resourcemanager.folders.get`` permission on
+    """
+    async with FoldersAsyncClient() as folders_client:
+        search_folders_pager = await folders_client.search_folders()
+        async for folders_page in search_folders_pager.pages:
+            yield parse_protobuf_messages(folders_page.folders)
+
+
+async def search_all_organizations() -> ASYNC_GENERATOR_RESYNC_TYPE:
+    """
+    Search for organizations that the caller has ``resourcemanager.organizations.get``` permission on
+    """
+    async with OrganizationsAsyncClient() as organizations_client:
+        search_organizations_pager = await organizations_client.search_organizations()
+        async for organizations_page in search_organizations_pager.pages:
+            yield parse_protobuf_messages(organizations_page.organizations)
+
+
+async def get_project(project_name: str) -> RAW_ITEM:
+    async with ProjectsAsyncClient() as projects_client:
+        return parse_protobuf_message(
+            await projects_client.get_project(name=project_name)
+        )
+
+
+async def get_folder(folder_name: str) -> RAW_ITEM:
+    async with FoldersAsyncClient() as folders_client:
+        return parse_protobuf_message(await folders_client.get_folder(name=folder_name))
+
+
+async def get_organization(organization_name: str) -> RAW_ITEM:
+    async with OrganizationsAsyncClient() as organizations_client:
+        return parse_protobuf_message(
+            await organizations_client.get_organization(name=organization_name)
+        )
+
+
+async def get_topic(topic_id: str) -> RAW_ITEM:
+    """
+    The Topics are handled specifically due to lacks of data in the asset itself within the asset inventory- e.g. some properties missing.
+    Here the PublisherAsyncClient is used, ignoring state in assets inventory
+    """
+    async with PublisherAsyncClient() as async_publisher_client:
+        return parse_protobuf_message(
+            await async_publisher_client.get_topic(topic=topic_id)
+        )
+
+
+class ResourceNotFoundError(Exception):
+    pass
+
+
+async def search_resource(
+    project_id: str, asset_kind: str, asset_name: str
+) -> RAW_ITEM:
+    try:
+        resource = [
+            resources
+            async for resources in search_all_resources_in_project(
+                project_id, asset_kind, asset_name
+            )
+        ][0][0]
+    except IndexError:
+        raise ResourceNotFoundError(
+            f"Found no asset named {asset_name} with type {asset_kind}"
+        )
```

Comment:
if you can fetch a spcific item then lets modify it a bit to use cool python stuff
```suggestion
        resource = anext(resources in search_all_resources_in_project(project_id, asset_kind, asset_name))[0]
    except StopIteration:
        raise ResourceNotFoundError(
            f"Found no asset named {asset_name} with type {asset_kind}"
        )
    except IndexError:
        raise ResourceNotFoundError(f"Got an empty page response when trying to Fetch asset {asset_name} with type {asset_kind}")
```

Reply:
This implementation didn't work, for now i will leave as is as i could find any prettier way to make this work